bakery/codec.go

```diff
@@ -9,49 +9,55 @@ import (
 
 	"golang.org/x/crypto/nacl/box"
 	"gopkg.in/errgo.v1"
+
+	"gopkg.in/macaroon-bakery.v2-unstable/bakery/checkers"
 )
 
-type caveatIdRecord struct {
+type caveatRecord struct {
 	RootKey   []byte
 	Condition string
 }
 
-// caveatId defines the format of a third party caveat id.
-type caveatId struct {
+// caveatId defines the format of a V1 JSON-encoded third party caveat id.
+type caveatJSON struct {
 	ThirdPartyPublicKey *PublicKey
 	FirstPartyPublicKey *PublicKey
 	Nonce               []byte
 	Id                  string
 }
 
-// encodeCaveatId creates a third-party caveat with the given condtion
+// encodeCaveat encrypts a third-party caveat with the given condtion
 // and root key. The thirdPartyInfo key holds information about the
 // third party we're encrypting the caveat for; the key is the
 // public/private key pair of the party that's adding the caveat.
 //
-// The caveat id will be encoded according to the version information
+// The caveat will be encoded according to the version information
 // found in thirdPartyInfo.
-func encodeCaveatId(
+func encodeCaveat(
 	condition string,
 	rootKey []byte,
 	thirdPartyInfo ThirdPartyInfo,
 	key *KeyPair,
+	ns *checkers.Namespace,
 ) ([]byte, error) {
 	switch thirdPartyInfo.Version {
 	case Version0, Version1:
-		return encodeCaveatIdV1(condition, rootKey, &thirdPartyInfo.PublicKey, key)
+		return encodeCaveatV1(condition, rootKey, &thirdPartyInfo.PublicKey, key)
+	case Version2:
+		return encodeCaveatV2(condition, rootKey, &thirdPartyInfo.PublicKey, key)
+	case Version3:
+		return encodeCaveatV3(condition, rootKey, &thirdPartyInfo.PublicKey, key, ns)
 	default:
-		// It's V2 or later - use V2.
-		return encodeCaveatIdV2(condition, rootKey, &thirdPartyInfo.PublicKey, key)
+		return nil, errgo.Newf("unknown bakery version %v", thirdPartyInfo.Version)
```

Comment:
I think I prefer the old behaviour here. shouldn't we use the latest version we know about in the case that the third party has a later version than us?